pretty/src/pane.rs

```diff
@@ -0,0 +1,300 @@
+                    child_pane.render(child_note, style, get_content.clone())?;
+                }
+            }
+            PaneNotation::Vert { panes, style } => {
+                let child_notes: Vec<_> = panes.iter().map(|p| &p.1).collect();
+                let child_sizes: Vec<_> = panes.iter().map(|p| p.0).collect();
+                let total_height = self.rect().height() as usize;
+                let heights: Vec<_> = divvy(total_height, &child_sizes)
+                    .ok_or(PaneError::ImpossibleDemands)?
+                    .into_iter()
+                    .map(|n| n as Row)
+                    .collect();
+                let style = style.or(parent_style);
+                for (rect, child_note) in self
+                    .rect()
+                    .vert_splits(&heights)
+                    .zip(child_notes.into_iter())
+                {
+                    let mut child_pane = self.sub_pane(rect).ok_or(PaneError::NotSubPane)?;
+                    child_pane.render(child_note, style, get_content.clone())?;
+                }
+            }
+            PaneNotation::Content { content, style } => {
+                // TODO how to use style?
+                let _style = style.or(parent_style).unwrap_or_default();
+                let width = self.rect().width();
+                let doc = get_content(content).ok_or(PaneError::Content)?;
+
+                // Put the top of the cursor at the top of the pane.
+                // TODO support fancier positioning options.
+                let cursor_region = doc.locate_cursor(width);
+                let doc_pos = Pos {
+                    col: 0,
+                    row: cursor_region.pos.row,
+                };
+                doc.pretty_print(width, self, doc_pos)?;
+            }
+            PaneNotation::Fill { ch, style } => {
+                let style = style.or(parent_style).unwrap_or_default();
+                let line: String = iter::repeat(ch)
+                    .take(self.rect().width() as usize)
+                    .collect();
+                let rows = self.rect().height();
+                for row in 0..rows {
+                    self.print(Pos { row, col: 0 }, &line, style)?;
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl PaneSize {
+    fn get_fixed(&self) -> Option<usize> {
+        match self {
+            PaneSize::Fixed(n) => Some(*n),
+            _ => None,
+        }
+    }
+
+    fn get_proportional(&self) -> Option<usize> {
+        match self {
+            PaneSize::Proportional(n) => Some(*n),
+            _ => None,
+        }
+    }
+}
+
+impl<T> From<T> for PaneError<T>
+where
+    T: fmt::Debug,
+{
+    fn from(e: T) -> PaneError<T> {
+        PaneError::PrettyWindow(e)
+    }
+}
+
+fn divvy(cookies: usize, demands: &[PaneSize]) -> Option<Vec<usize>> {
+    let total_fixed: usize = demands.iter().filter_map(|demand| demand.get_fixed()).sum();
```

Comment:
TODO: make `divvy` generic over int types

Reply:
and replace the two `.map()`s

Reply:
Turns out that's way more complicated than the `map`s.